core/src/main/java/org/infinispan/util/concurrent/locks/containers/AbstractPerEntryLockContainer.java

```diff
@@ -137,4 +114,36 @@ public String toString() {
             "locks=" + locks +
             '}';
    }
+
+   private class Locker implements EquivalentConcurrentHashMapV8.BiFun<Object, L, L> {
```

Comment:
I could not get results precise enough to be able to compare these two implementations - in my benchmarks, Infinispan was sometimes suddenly faster for a brief period of time without any obvious reason, spoiling the results.
However, by merging the function and value returned by reference, you allocate only one object instead of two. I can't tell whether this would be eliminated by compiler anyway.